auto_submit/lib/configuration/repository_configuration_manager.dart

```diff
@@ -0,0 +1,174 @@
+import 'package:auto_submit/service/github_service.dart';
+import 'package:auto_submit/service/log.dart';
+import 'package:github/github.dart';
+import 'package:mutex/mutex.dart';
+import 'package:neat_cache/neat_cache.dart';
+
+/// The [RepositoryConfigurationManager] is responsible for fetching and merging
+/// the autosubmit configuration from the Org level repository and if needed
+/// fetching the override configuration from the pull request repository.
+///
+/// It will attempt to access the cache first before repulling the configuraiton
+/// from the repositories. This is currently set at a 10 minute TTL.
+class RepositoryConfigurationManager {
+  final Mutex _mutex = Mutex();
+
+  static const String fileSeparator = '/';
+  // This is the well named organization level repository and configuration file
+  // we will read before looking to see if there is a local file with
+  // overwrites.
+  static const String orgRepository = '.github';
+  static const String dirName = 'autosubmit';
+  static const String fileName = 'autosubmit.yml';
+
+  final Config config;
+  final Cache cache;
+
+  RepositoryConfigurationManager(this.config, this.cache);
+
+  /// Read the configuration from the cache given the slug, if the config is not
+  /// in the cache then go and get it from the repository and store it in the
+  /// cache.
+  ///
+  /// Entries will be stored in the cache as config/slug/autosubmit.yaml
+  Future<RepositoryConfiguration> readRepositoryConfiguration(
+    RepositorySlug slug,
+  ) async {
+    await _mutex.acquire();
+    try {
+      // Get the contents from the cache or go to github.
+      final cacheValue = await cache['${slug.fullName}$fileSeparator$fileName'].get(
+        () async => _getConfiguration(slug),
+        const Duration(minutes: 10),
+      );
+      // final String yamlContents = String.fromCharCodes(cacheValue!);
+      final String cacheYaml = String.fromCharCodes(cacheValue);
+      log.info('Converting yaml to RepositoryConfiguration: $cacheYaml');
+      return RepositoryConfiguration.fromYaml(cacheYaml);
+    } finally {
+      _mutex.release();
+    }
+  }
+
+  /// Collect the configuration from github and handle the cache conversion to
+  /// bytes.
+  Future<List<int>> _getConfiguration(
+    RepositorySlug slug,
+  ) async {
+    // Read the org level configuraiton file first.
+    log.info('Getting org level configuration.');
+    // Get the Org level configuration.
+    final RepositorySlug orgSlug = RepositorySlug(slug.owner, orgRepository);
+    final GithubService githubService = await config.createGithubService(orgSlug);
+    final String orgLevelConfig = await githubService.getFileContents(orgSlug, '$dirName$fileSeparator$fileName');
+    final RepositoryConfiguration globalRepositoryConfiguration = RepositoryConfiguration.fromYaml(orgLevelConfig);
+
+    // Collect the default branch if it was not supplied.
+    if (globalRepositoryConfiguration.defaultBranch!.isEmpty) {
```

Comment:
Do we allow the configuration to contradict github in terms of default branch? or what is the use case for globalRepositoryConfiguration.defaultBranch?

Reply:
Hmm, this may be from a convenience for myself from early testing when I was assuming that default branch was being supplied to us. Let me refactor this.